Implemented CAST from VARBINARY to DATETIME in Babelfish.

Yvinayak07 opened the pull request:
### Description

Previously, Babelfish did not support casting from VARBINARY to DATETIME. We have now implemented this functionality to align with T-SQL's capabilities. This new feature allows users to cast VARBINARY data types to DATETIME.

cherry-picked from https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3481.

Task: BABEL-3070
Signed-off-by: yashneet vinayak [yashneet@amazon.com](mailto:yashneet@amazon.com)

### Test Scenarios Covered ###
* **Use case based -**Yes


* **Boundary conditions -**Yes


* **Arbitrary inputs -**Yes


* **Negative test cases -**Yes


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**Yes



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Implemented CAST from VARBINARY to DATETIME in Babelfish

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Anikait143 commented on contrib/babelfishpg_common/sql/datetime.sql:
```diff
@@ -312,6 +312,13 @@ LANGUAGE C VOLATILE STRICT PARALLEL SAFE;
+LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE CAST (sys.BBF_VARBINARY AS DATETIME)
+WITH FUNCTION sys.varbinary2datetime(sys.BBF_VARBINARY) AS ASSIGNMENT;
```
This Cast should be defined `IMPLICIT`

Yvinayak07 replied:
Yes as per TSQL doc this cast should be `IMPLICIT`. Missed this will correct.

Anikait143 commented on contrib/babelfishpg_common/sql/datetime.sql:
```diff
@@ -312,6 +312,13 @@ LANGUAGE C VOLATILE STRICT PARALLEL SAFE;
+CREATE OR REPLACE FUNCTION sys.varbinary2datetime(sys.BBF_VARBINARY)
+RETURNS DATETIME
+AS 'babelfishpg_common', 'varbinary_datetime'
+LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
```
Can we check if this cast depends on external factors such as timezone setting. If so, then we should define this CAST as `STABLE`

Yvinayak07 replied:
Tested usingdifferent timezone it produces same result.

Anikait143 commented on contrib/babelfishpg_common/sql/upgrades/babelfish_common_helper--5.0.0--5.1.0.sql:
```diff
@@ -127,5 +127,13 @@ CREATE OR REPLACE AGGREGATE sys.max(sys.money) (
+AS 'babelfishpg_common', 'varbinary_datetime'
+LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE CAST (sys.BBF_VARBINARY AS DATETIME)
+WITH FUNCTION sys.varbinary2datetime(sys.BBF_VARBINARY) AS ASSIGNMENT;
```
We should check if there is already existing CAST present to avoid upgrade script exceptions

Yvinayak07 replied:
Yes will do that once the dev branch is open for xerus as we need to add this to 5.1.0 to 5.2.0 path.

Anikait143 commented on test/JDBC/upgrade/16_7/schedule:
```diff
@@ -139,7 +139,7 @@ BABEL-CHECK-CONSTRAINT
 BABEL-CROSS-DB
 babel_datatype_sqlvariant
-babel_datetime
+babel_datetime-before-17_3
```
If we are merging this in older majors as well, then we should start with them first so that we can avoid changing this later.

Yvinayak07 replied:
Raised Pr for lower version once this is approved will cherry-pick all changes their and get them merged.

Anikait143 commented on test/JDBC/input/babel_datetime-vu-prepare.sql:
Let's also include functions and views to test dependency and implicit casting

Anikait143 commented on test/JDBC/expected/babel_datetime-vu-verify.out:
```diff
@@ -21,6 +21,12 @@ datetime
 1900-01-01 23:40:30.0
 1900-01-01 23:40:30.0
 2020-03-14 00:00:00.0
+1900-01-01 00:02:30.3
+2023-06-15 14:30:45.12
+2000-02-29 12:00:00.0
+1753-01-01 00:00:00.0
+2023-06-15 14:30:45.123
+9999-12-31 23:59:59.997
```
Have we verified these with TSQL for precision?

Yvinayak07 replied:
Looks like JDBC issue.
Checked packets in wireshark looks same for jdbc and odbc (but still output diff).

Anikait143 commented on contrib/babelfishpg_common/sql/datetime.sql:
```diff
@@ -312,6 +312,13 @@ LANGUAGE C VOLATILE STRICT PARALLEL SAFE;
+AS 'babelfishpg_common', 'varbinary_datetime'
+LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE CAST (sys.BBF_VARBINARY AS DATETIME)
```
What will happen if the hex is of Binary datatype?

Yvinayak07 replied:
It will give error as we don't support cast from binary to datetime.

sumitj824 commented on contrib/babelfishpg_common/sql/datetime.sql:
```diff
@@ -312,6 +312,13 @@ LANGUAGE C VOLATILE STRICT PARALLEL SAFE;
+LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE CAST (sys.BBF_VARBINARY AS DATETIME)
+WITH FUNCTION sys.varbinary2datetime(sys.BBF_VARBINARY) AS ASSIGNMENT;
```
Why are we marking this as  `ASSIGNMENT` instead of `IMPLICIT` cast?

Yvinayak07 replied:
Yes as per TSQL doc this cast should be `IMPLICIT`. Missed this will correct.

sumitj824 commented on contrib/babelfishpg_common/src/datetime.c:
```diff
@@ -764,6 +765,69 @@ timestamp_datetime(PG_FUNCTION_ARGS)
+	unsigned char		*buffer;
+
+	/* TSQL datetime is 8 bytes */
+	if (size != sizeof(int64) && size != sizeof(int32))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid binary size for datetime conversion")));
```
Do you have added testcases for these cases?

Yvinayak07 replied:
Yes we have already added test case for this condition.
https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3448/files#diff-f3d30b71ec76a7f337c767d197028b41b676ac980f76808eaa6373989559ca5fR514

sumitj824 commented on contrib/babelfishpg_common/src/datetime.c:
```diff
@@ -764,6 +765,69 @@ timestamp_datetime(PG_FUNCTION_ARGS)
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid binary size for datetime conversion")));
+
+	buffer = (unsigned char *)palloc(size);
+	memcpy(buffer, data, size);
```
Can you explain the reason behind this? Why can't we directly use the data?

Yvinayak07 replied:
yes, agreed data seems to be redundant here, we can use buffer instead.

sumitj824 commented on contrib/babelfishpg_common/src/datetime.c:
```diff
@@ -764,6 +765,69 @@ timestamp_datetime(PG_FUNCTION_ARGS)
+	}
+	else
+	{
+		days = (buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];
+		time_part = (buffer[4] << 24) | (buffer[5] << 16) | (buffer[6] << 8) | buffer[7];
+	}
+
+	/* Convert time_part to microseconds */
+	ms_value = ((int64)time_part * 10LL) / 3LL;
+	usecs = ms_value * 1000;
+
+	if (days < 0)
+	{
+		/* Handle pre-1900 dates */
+		int64 day_value = (int64) (((int64) days) & ((int64) 0xFFFFFFFF));
+		int64 total_usecs = (day_value - 0xFFFF2E46LL) * USECS_PER_DAY + usecs;
+		result = MIN_DATETIME + total_usecs;
+	}
+	else
+	{
+		/* Handle post-1900 dates */
+		int64 total_usecs = days * USECS_PER_DAY + usecs;
+		result = TSQL_DEFAULT_DATETIME + total_usecs;
+	}
```
I think we should use the existing APIs for this calculation.

Yvinayak07 replied:
We need two cases to handle where API is just calculating days from default date where the return type is float.

sumitj824 commented on test/JDBC/input/babel_datetime-vu-verify.sql:
```diff
@@ -139,3 +139,181 @@ select pg_typeof(c1) FROM (SELECT '2016-12-26 23:30:05'::datetime as C1 UNION SE
 select pg_typeof(c1) FROM (SELECT '2016-12-26 23:30:05'::datetime as C1 UNION SELECT '2016-12-26'::date as C1) T
 go
+
+-- Test Varbinary to Datetime conversion using CAST
```
Please add more negative testcases.

Yvinayak07 replied:
Added more test cases.

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-3070

Commit:
pulling latest pull

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
adding schedule file in 14_7

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
updating schedule 15_12

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
changing name of before file according to correct version

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
changing before file name in jdbc_schedule ignore

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
adding exception handling in upgrade script

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Yvinayak07 committed:
Implemented CAST from VARBINARY to DATETIME in Babelfish. (#3481)

Previously, Babelfish did not support casting from VARBINARY to DATETIME. We have now implemented this functionality to align with T-SQL's capabilities. This new feature allows users to cast VARBINARY data types to DATETIME. We have also added support for convert functions from VARBINARY to DATETIME and string datatypes to DATETIME.

Task: BABEL-3070
Signed-off-by: yashneet vinayak yashneet@amazon.com

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-3070

Commit:
correct logic after merged in 3_X

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
correcting schedule files

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
depricating babelfish_conv_helper_to_datetime functions

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
correcting syntax for deprication

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
correcting syntax

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
adding exception handling

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
depricating anyelement

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
Revert "depricating anyelement"

This reverts commit 25368d20d92dbcb8f58f90448f6d20d8e9c286fc.

Commit:
deleting depricating objects

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
Revert "deleting depricating objects"

This reverts commit cca8168d8e112897848acb8624237d2b556ba96b.

Commit:
disabling test

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
removing redundant changes

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
aMVU testing

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
MVU testing

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
Revert "MVU testing"

This reverts commit 311cfca07d18cdc5a5ac8bf89b1d7ed952e2ceb6.

Commit:
Revert "aMVU testing"

This reverts commit efe6d33a2e7d29b2db9d1f69f45f8499433b395c.

Commit:
reverting change in github actions

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
correcting 15_13 schedule file

Signed-off-by: yashneet vinayak <yashneet@amazon.com>

Commit:
empty commit

Signed-off-by: yashneet vinayak <yashneet@amazon.com>
